doc/adrs/057_add_support_for_tool_precondition.adoc

```diff
@@ -0,0 +1,31 @@
+= ADR-057 - Add support for tool preconditions
+
+== Context
+
+As of now, to check if a tool can be applied on a diagram element (as a source or target of the tool), we check that its diagram description is within the candidates of the tool.
+The semantic object represented by the diagram element does not play a role in this checking process.
+Only when the tool is applied do we know if the tool will successfully apply to the semantic object.
+
+== Decision
+
+We will add an way for a tool to accept a semantic object as a source or target in the diagram tools interfaces `SingleClickOnTwoDiagramElementsTool` and `SingleClickOnDiagramElementTool`.
+The connector tool will check against the backend if the tools it is given are valid for the chosen source and target.
```

Comment:
This would allow the frontend to send one request to ask for the relevant tools, something like this:

```
query getConnectorTools($editingContextId: ID!, representationId: ID!, $sourceDiagramElementId: ID!, $targetDiagramElementId: ID!): {
  viewer {
    editingContext(editingContextId: $editingContextId) {
      representation(representationId: $representationId) {
        description {
          ... on DiagramDescription {
            connectorTools(sourceDiagramElementId: $sourceDiagramElementId, targetDiagramElementId: $targetDiagramElementId) {
              id
              label
              imageURL
            }
          }
        }
      }
    }
  }
}
```

And then, the backend could do anything it want to compute said tools, that's irrelevant for the frontend.

Reply:
When the palette is loaded after clicking on the source element, all available tools are already at the frontend, they are given to the ContextualMenu, since we already have the tools all that remains to do is validate them against the target that may be selected later.

Is it preferrable to ask again for the tools ( and validate them in the backend before returning ) when ContextualMenu is displayed instead of asking the backend if the tools given as a prop are valid for the given source and target ?

Reply:
We should simply ask for the minimum of information necessary to perform the job on the frontend. By only asking for the relevant tools in the `ContextualMenu`, we gain several advantages:

- The code is simpler (no need to pass data from components to components)
- We may end up removing code elsewhere
- We can end up displaying tools in the `ContextualMenu` that may not be available elsewhere
- We don't need to introduce anything related to preconditions in the API so we get a simpler lifecycle

When we need to display the tools that should be available, we ask for them. How are they computed? The frontend does not need to care. We just need an interface on the backend, similar in philosophy to the `IToolSectionsProvider` to answer the question "what tools should be displayed to interact with both this source and target?".

Some implementation of this interface may end up leveraging preconditions from an odesign file like `CompatibilityToolSectionsProvider` does but that's.a concern for the person creating the diagram description. As you can see with this code (starting from the `GetToolSectionsEventHandler`) we do not have any concept of precondition in Sirius Components (just in the Sirius Desktop compatibility layer since Sirius Desktop has this concept). We don't need it.